typed/union_test.go

```diff
@@ -0,0 +1,224 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package typed_test
+
+import (
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/typed"
+)
+
+var unionParser = func() typed.ParseableType {
+	parser, err := typed.NewParser(`types:
+- name: union
+  struct:
+    fields:
+    - name: discriminator
+      type:
+        scalar: string
+    - name: a
+      type:
+        scalar: numeric
+    - name: b
+      type:
+        scalar: numeric
+    - name: c
+      type:
+        scalar: numeric
+    - name: one
+      type:
+        scalar: numeric
+    - name: two
+      type:
+        scalar: numeric
+    - name: three
+      type:
+        scalar: numeric
+    unions:
+      - discriminator: discriminator
+        fields: [a, b, c]
+      - fields: [one, two, three]`)
```

Comment:
Test looks good, maybe we should also try some nested structs

Reply:
Also we should test that creating a new object should still set the discriminator correctly